go/vt/vtgate/engine/route_warming_reads_test.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/srvtopo"
+	"vitess.io/vitess/go/vt/vtgate/evalengine"
+	"vitess.io/vitess/go/vt/vtgate/vindexes"
+
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+type warmingReadsVCursor struct {
+	*loggingVCursor
+	warmingReadsPercent     int
+	warmingReadsChannel     chan bool
+	warmingReadsExecuteFunc func(context.Context, Primitive, []*srvtopo.ResolvedShard, []*querypb.BoundQuery, bool, bool)
+}
+
+func (vc *warmingReadsVCursor) GetWarmingReadsPercent() int {
+	return vc.warmingReadsPercent
+}
+
+func (vc *warmingReadsVCursor) GetWarmingReadsChannel() chan bool {
+	return vc.warmingReadsChannel
+}
+
+func (vc *warmingReadsVCursor) CloneForReplicaWarming(ctx context.Context) VCursor {
+	clone := &warmingReadsVCursor{
+		loggingVCursor:          vc.loggingVCursor,
+		warmingReadsPercent:     vc.warmingReadsPercent,
+		warmingReadsChannel:     vc.warmingReadsChannel,
+		warmingReadsExecuteFunc: vc.warmingReadsExecuteFunc,
+	}
+	clone.onExecuteMultiShardFn = vc.warmingReadsExecuteFunc
+	return clone
+}
+
+func TestWarmingReadsSkipsForUpdate(t *testing.T) {
+	vindex, _ := vindexes.CreateVindex("hash", "", nil)
+	testCases := []struct {
+		name       string
+		query      string
+		shouldSkip bool
+	}{
+		{
+			name:       "SELECT FOR UPDATE",
+			query:      "SELECT * FROM users WHERE id = 1 FOR UPDATE",
+			shouldSkip: true,
+		},
+		{
+			name:       "SELECT FOR UPDATE mixed case",
+			query:      "SELECT * FROM users WHERE id = 1 FoR UpDaTe",
```

Comment:
If there's two spaces or a newline or a comment in between for update, this will not match correctly.